src/pages/githubAssessments/GitHubMissionListing.tsx

```diff
@@ -0,0 +1,303 @@
+
+      const handleChange = (values: React.ReactNode[]) => {
+        setValues(values);
+      };
+
+      const tagFilter = (
+        <TagInput
+          leftIcon={IconNames.FILTER}
+          onChange={handleChange}
+          placeholder="Separate tags with commas..."
+          rightElement={clearButton}
+          tagProps={{ minimal: true }}
+          values={values}
+        />
+      );
+
+      const cards =
+        values.length === 0
+          ? browsableMissions.map(element => convertMissionToCard(element, isMobileBreakpoint))
+          : filteredMissions.map(element => convertMissionToCard(element, isMobileBreakpoint));
+
+      setDisplay(
+        <>
+          {tagFilter}
+          <Divider />
+          {cards}
+        </>
+      );
+    }
+  }, [browsableMissions, isMobileBreakpoint, setDisplay, values]);
+
+  return (
+    <div className="Academy">
+      <div className="Assessment">
+        <ContentDisplay display={display} loadContentDispatch={getGitHubOctokitInstance} />
+      </div>
+    </div>
+  );
+};
+
+/**
+ * Retrieves BrowsableMissions information from a mission repositories and sets them in the page's state.
+ *
+ * @param octokit The Octokit instance for the authenticated user
+ * @param setBrowsableMissions The React setter function for an array of BrowsableMissions
+ * @param setDisplay The React setter function for the page's display
+ */
+async function retrieveBrowsableMissions(
+  octokit: Octokit,
+  setBrowsableMissions: (browsableMissions: BrowsableMission[]) => void,
+  setDisplay: (display: JSX.Element) => void
+) {
+  if (octokit === undefined) return;
+
+  setDisplay(
+    <NonIdealState description="Loading Missions" icon={<Spinner size={Spinner.SIZE_LARGE} />} />
+  );
+
+  const allRepos = (await octokit.repos.listForAuthenticatedUser({ per_page: 100 })).data;
+  const correctlyNamedRepos = allRepos.filter((repo: any) => repo.name.startsWith('sa-'));
+  const foundMissionRepos: MissionRepoData[] = [];
+
+  for (let i = 0; i < correctlyNamedRepos.length; i++) {
+    const repo = correctlyNamedRepos[i] as any;
+
+    const files = (
+      await octokit.repos.getContent({
+        owner: repo.owner.login,
+        repo: repo.name,
+        path: ''
+      })
+    ).data;
+
+    if (!Array.isArray(files)) {
+      setDisplay(<NonIdealState title="There are no assessments." icon={IconNames.FLAME} />);
+      return;
+    }
+
+    if (files.find(file => file.name === '.metadata') !== undefined) {
+      const missionRepoData: MissionRepoData = {
+        repoOwner: repo.owner.login,
+        repoName: repo.name,
+        dateOfCreation: new Date(repo.created_at)
+      };
+      foundMissionRepos.push(missionRepoData);
+    }
+  }
+
+  const browsableMissions: BrowsableMission[] = [];
+
+  for (let i = 0; i < foundMissionRepos.length; i++) {
+    browsableMissions.push(await convertRepoToBrowsableMission(foundMissionRepos[i], octokit));
+  }
```

Comment:
File contents could not be retrieved without making extra API calls. Code rewritten to use `Promise.all`:
https://github.com/chekjun/cadet-frontend/blob/8c389af6cb89646e2feb3eed084fbe405c06468f/src/pages/githubAssessments/GitHubMissionListing.tsx#L242-L250

Reply:
if im not wrong, this is still synchronous, since `convertRepoToBrowsableMission` contains an `await` which blocks at each iteration of map. is it possible to add more missions to the github classroom? then we can test this proper
```
const missionPromises = foundMissionRepos.map(missionRepoData => convertRepoToBrowsableMission(missionRepoData, octokit));
```

Reply:
How should I rewrite this code? I've tried a few things but they appear to still be asynchronous.